src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import data.fintype.basic
+import algebra.big_operators.basic
+import data.equiv.fin
+
+/-!
+# The Hales-Jewett theorem
+
+We prove the Hales-Jewett theorem and deduce Van der Waerden's theorem as a corollary.
+
+We consistently work with finite types instead of natural numbers. That is, we write `m, n, k` for
+(finite) types where one might traditionally take `m, n, k` to be natural numbers. This allows us to
+work directly with `m`, `option m`, `(n → m) → k`, and `n ⊕ n'` instead of `fin m`, `fin (m+1)`,
+`fin (k^(m^n))`, and `fin (n + n')`.
+
+We take '`X` is monochromatic' to mean 'there exists a color `c` such that every point of `X` has
+the color `c`' as opposed to 'any `x y : X` have the same color' since the former is generally
+easier to work with. This means that our formulation of Hales-Jewett would be false when `m` and `k`
+are both empty (then there can be a coloring even though there is no color). Thus we require `m` to
+be nonempty in the statement of `exists_mono_in_high_dimension`.
+
+We do not prove the finitary version of Van der Waerden's theorem (which is equivalent to the
+infinite one by compactness), although it follows from the same proof.
+
+In principle our proof should give computable upper bounds on the numbers `HJ(m, k)`, but this would
+require rewriting the proof to avoid `Prop`, and the upper bounds are enormous anyway.
+
+## Main results
+
+- `combinatorics.line.exists_mono_in_high_dimension m k`: the Hales-Jewett theorem.
+- `combinatorics.exists_mono_homothetic_copy M S f k C`: a generalization of Van der Waerden's
+theorem to an arbitrary commutative monoid. The proof is a simple application of Hales-Jewett.
+
+## Tags
+
+combinatorial line, Ramsey theory, arithmetic progession
+
+### References
+
+* https://en.wikipedia.org/wiki/Hales%E2%80%93Jewett_theorem
+
+Note that this file does not use the same notation as Wikipedia.
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+instance (n : pnat) : nonempty (fin n) := ⟨⟨0, n.pos⟩⟩
+
+/-- An induction principle for types analogous to `pnat.rec_on`. To prove a statement about
+all nonempty finite types, it suffices to consider `unit` and `option α` where we know the
+statement for `α`, provided that our statement is preserved by `≃`.-/
+lemma fintype_induction {P : Type → Prop} (of_equiv : ∀ {α β}, α ≃ β → P α → P β)
+  (h_unit : P unit) (h_option : ∀ {α} [fintype α] [nonempty α], P α → P (option α))
+  (α) [fintype α] [nonempty α] : P α :=
+begin
+  suffices : ∀ n : pnat, P (fin n),
+  { refine of_equiv (fintype.equiv_fin α).symm
+      (this ⟨fintype.card α, finset.card_pos.mpr finset.univ_nonempty⟩), },
+  intro n, apply n.rec_on,
+  exact of_equiv fin_one_equiv.symm h_unit,
+  exact λ _ ih, of_equiv (fin_succ_equiv _).symm (h_option ih),
+end
```

Comment:
This is a nice induction principle especially for combinatorics proofs like this one - perhaps it belongs elsewhere?

Reply:
Possibly, but I'm not sure where

Reply:
I think the current induction pattern is specialized enough that it might not be very useful in other places. There is a more natural pattern that's just analogous to `nat.rec`: "any finite type is `empty` or `option α`". The current pattern is like this one, plus a case distinction on whether `α` is empty or not, plus the identification `option empty ≃ unit`...

Reply:
Ok, I'll probably open a new PR for this.

Reply:
This is now in #8158. I think the natural `empty`/`option α` pattern will do, we can easily get the specialized version from it.